examples/matrix-factorisation/SparseRatingMatrix.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Main where
+
+import Control.Monad (foldM, when)
+import MF
+import Torch.Autograd (grad, makeIndependent, toDependent)
+import Torch.DType (DType (Float))
+import Torch.Functional (matmul)
+import Torch.Internal.Managed.Type.Context (manual_seed_L)
+import Torch.NN
+  ( Randomizable,
+    sample,
+    sgd,
+  )
+import Torch.Tensor (Tensor, asTensor, toDouble, toInt)
+import Torch.TensorFactories (randint', randn')
+
+type Item = Int
+
+type User = Int
+
+type RatingValue = Double
+
+data Rating = Rating Item User RatingValue deriving (Show)
+
+data Ratings = Ratings [Rating]
+
+data RatingsSpec
+  = RatingsSpec {usersNumber :: Int, itemsNumber :: Int, ratingNumber :: Int}
+
+instance Randomizable RatingsSpec Ratings where
+  sample (RatingsSpec n m l) = do
+    ratings <- randn' [1, l]
+    users <- randint' 1 n [1, l]
+    items <- randint' 1 m [1, l]
+    let list_ratings = toDouble <$> (getEntryByIndex ratings 0) <$> [0 .. (l -1)]
+        list_users = toInt <$> (getEntryByIndex users 0) <$> [0 .. (l -1)]
+        list_items = toInt <$> (getEntryByIndex items 0) <$> [0 .. (l -1)]
+    pure $ Ratings [Rating u i r | (u, i, r) <- zip3 list_users list_items list_ratings]
+
+type TensorEntries = Double
+
+lossMF :: MatrixFact -> [Rating] -> Tensor
+lossMF (MatrixFact u v) rating_list =
+  let uv = mulMF (MatrixFact u v)
```

Comment:
you are computing the full matrix `uv` here. that may not always be possible due to its size.
you want a way to only compute the elements of `uv` that correspond to `users` and `items` in `RatingBatch`.

Reply:
I realize that this may not be easy.
in pytorch, there's https://pytorch.org/docs/stable/torch.html#torch.index_select.
I *think* the equivalent in hasktorch is https://github.com/hasktorch/hasktorch/blob/master/hasktorch/src/Torch/Functional/Internal.hs#L336, but I'm not exactly sure.

Reply:
once you have that, you can use that
```latex
R_{ij} = \sum_l U_{il} V{lj}
```
that is, you'll have a bunch of vector multiplications.

Reply:
that can be written as a matrix multiplication as well, I just realize.
just use index select on `u` and `v` with `users` and `items`, respectively, and then `matmul` those together. that will give you the truncated `R` you need I think.
then you reshape it to a 1-dim tensor of length `batchSize` which you can feed into `mse_loss`.

Reply:
never mind, that's not the case ^^^

Reply:
run a batched vector multiplication, `bmv`, instead.

Reply:
It is a great point. I tried to change the code such that we only multiply the rows and columns of `u,` and `v` when it is needed.